### ISSUE TITLE
Point to Point Chapter RC Review

### ISSUE THREAD

**Issue by wesbland** [mpi-forum/mpi-issues#336]
# Problem

### Section 3.2.3

- [x] Advice to users at the end:

  I know this part is a brief summary about communicators, but reading that (basically) COMM_WORLD is enough looked strange to me whenever I think about collectives on a subset of processes.

### Section 3.2.5
- [x] In the definition of MPI_GET_COUNT line 2, is it normal to have "Status" with a capital in parenthesis at the end of "IN status" line?
  Many other function definitions seem to have "Status" with a capital in the document, while other things like "choice", "integer", "handle" have no starting capital.

There are some linebreaks that don't seem properly placed, causing the line to look too short:

### Section 3.2.2
- [x] Between "and" and "MPI_REAL16"
- [x] Between "REAL*4," and "REAL*8"
- [x] Between "equivalents" and "INTEGER" next page

### Section 3.2.5
- [x] Line 19, between "MPI_SOURCE" and "MPI_TAG"

## More from @devreal 

- [x] Page 109:
  The functions MPI_WAIT and MPI_TEST can be used to complete both sends and receives.
  -> add nonblocking/persistent collectives here?

- [x] Page 114 (MPI_WAITALL, also applies to MPI_TESTALL)
  lists length -> list length (for consistency with MPI_WAITANY and MPI_STARTALL)

- [x] Page 115 (MPI_TESTALL, also applies to MPI_IPROBE and MPI_TEST_CANCELLED)
  The flag argument is missing a description

- [x] Page 116 (MPI_WAITSOME)
  Returns in the first outcount locations of the array array_of_status the status for these completed operations
  -> array_of_status -> array_of_statuses

- [x] Page 121:
  Oxford comma: source, tag and length -> source, tag, and length

## More from @GuillaumeMercier 

- [x] pp. 92, 93 and 94
  * Adding "and binds to it all arguments of a send operation." to the description of MPI_B|S|Rsend_init to be consistent with MPI_Send_init?

- [x] p. 95:
  * "The arguments, request, is a handle" => remove both commas?
  * If the resquest is for a send with read mode" => "for a send operation" ?

- [x] p. 96:
  * Description of MPI_Startall "Start all communications" => "Starts all communications" ?
  * "The request becomes inactive" => "The associated request becomes inactive" ?

**Comment by wesbland** [comment]
I added more comments from @devreal

**Comment by raffenet** [comment]
- pg 108
Advice to Implementors states that "a call to MPI_WAIT should block only the calling thread". My understanding is the requirement is stronger, i.e. should -> shall.
- pg 109
both sends and receives -> all types of communication requests? (coll, rma, partitioned, ...)
- pg 110
Advice to users states that an error on a communication request that has been freed by the user "must be treated as fatal". Is this still true given the new default error handler (MPI_ERRORS_ABORT)?
- pg 111, ln 34 should -> shall?

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@raffenet

"MPI_WAIT should block only the calling thread" - I think this is only "should" because it is shall for a thread compliant implementation but thread compliance is not mandatory. Also, this is only advice - shall here would not be a normative requirement.

"both sends and receives -> all types of communication requests" - your suggestion is correct, but is it a chapter committee level change?

"must be treated as fatal" - I think would be a good thing to clarify but it is not a chapter committee level change, IMHO. The FT WG would need to weigh in on it, for a start.

"ln 34 should -> shall?" - the next sentence uses "will" but includes the additional reason that differentiates it from "should" --- namely, "if not satisfied by another [thing]". This is a very sensitive semantic. I do not wish to change should to shall without a vote. I expect such a vote to pass with no problems, but I've expected that before and been ... surprised.

**Comment by regrant** [comment]
## From @devreal (some of the comments ended up in chapter 4's review ticket by mistake)

- [ ] Page 130 (first paragraph of Section 3.9):
  > In such a situation, it may be possible to optimize the communication by binding the list of communication arguments to a persistent communication request once and, then, repeatedly using the request to initiate and complete operations.

  -> the commas around `then` seem unnecessary so I suggest removing both of them

- [ ] Page 130:
  collective communication persistent operations ->  persistent collective communication operations

- [ ] Page 130:
  These point-to-point persistent calls involve no communication.
  -> These point-to-point persistent procedures are local.

- [ ] Page 135:
  > The argument, request, is a handle...

  -> The Fortran call has two arguments so the `request` shouldn't be put into a subclause but instead just be named: The argument request is a handle...

- [ ] Page 135, same sentence:
  > The argument, request, is a handle returned by one of the previous five calls.

  -> Also reference Sections 6.13 (persistent collectives) and 8.8 (persistent neighborhood collectives)?

- [ ] Page 136:
  executed for i=0 ,. . ., count-1,

  -> executed for i=0, . . ., count-1, (notice the spacing around the first comma)

- [ ] Page 136:
  There is a paragraph break on lines 15-16 that seems unncessary (and confusing)

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@bgoglin 

Section 3.2.3 - the advice is strange in the context of Sessions as well. I do not propose to change this without a vote. My preference would be to remove the advice entirely.

Section 3.2.5 - status was capitalised in MPI-3.1 in this location. I agree it looks odd and may be difficult to justify. It is the only explicit struct parameter type in MPI, AFAIK. It is an exception, but is that a good justification?

**Comment by raffenet** [comment]
> @raffenet
> 
> "MPI_WAIT should block only the calling thread" - I think this is only "should" because it is shall for a thread compliant implementation but thread compliance is not mandatory. Also, this is only advice - shall here would not be a normative requirement.

👍 

> "must be treated as fatal" - I think would be a good thing to clarify but it is not a chapter committee level change, IMHO. The FT WG would need to weigh in on it, for a start.

cc: @wesbland @abouteiller

> 
> "ln 34 should -> shall?" - the next sentence uses "will" but includes the additional reason that differentiates it from "should" --- namely, "if not satisfied by another [thing]". This is a very sensitive semantic. I do not wish to change should to shall without a vote. I expect such a vote to pass with no problems, but I've expected that before and been ... surprised.

👍

**Comment by dholmes-epcc-ed-ac-uk** [comment]
Line breaks:

§3.2.2 page 72 (PDF) 32 (printed) line 43: "Between "and" and "MPI_REAL16" --- this one is maddening. I have no idea why Latex refuses to the do right thing here. I can force the REAL16 and the word "for" onto line 43 with a manual line break but then we get a jagged right edge for that line only. Even using non-breaking spaces fore and aft is ignored by Latex. @wgropp any idea why Latex is applying a terrible penalty/policy here?

Note that the other two for this §3.2.2 are dependent on the first - the rest of the paragraph changes (in sub-optimal ways) when the first one is "fixed".

**Comment by wesbland** [comment]
> "must be treated as fatal" - I think would be a good thing to clarify but it is not a chapter committee level change, IMHO. The FT WG would need to weigh in on it, for a start.

Yes, this should probably say something about using the default error handler, i.e., the one set for `MPI_COMM_SELF` or the default for the session.

**Comment by dholmes-epcc-ed-ac-uk** [comment]
Line breaks:

§3.2.5 --- I'm pretty sure that "status.Source" does not fit in the remaining space on the previous line. We could re-order the items in the list to put something shorter first?

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@wesbland must be fatal -> use default error handler seems like an errata to me.

**Comment by wesbland** [comment]
> @wesbland must be fatal -> use default error handler seems like an errata to me.

I think you're right. We should split this out to another issue to track it separately.

**Comment by RolfRabenseifner** [comment]
> @bgoglin
> Section 3.2.5 - status was capitalised in MPI-3.1 in this location. I agree it looks odd and may be difficult to justify. 

The subfields of status are capitalised MPI_TAG, MPI_ERROR, ... since MPI 1.1. Any change would not be backward compatible back to MPI-1.1.
In Fortran since MPI-3.0, we now provide the same structure with the same fields and use also capitalised names (although Fortran is not case sensitive).

> It is the only explicit struct parameter type in MPI, AFAIK. It is an exception, but is that a good justification?

No, in the Fortran mpi_f08 module (and additional for mixed codes also in the mpi module), all handle types are provided 
(also provided) as a structure with exactly one field: MPI_VAL.

**Comment by dholmes-epcc-ed-ac-uk** [comment]
"lists" -> "list" (x2) fixed by https://github.com/mpi-forum/mpi-standard/pull/361

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@RolfRabenseifner the question is about "(Status)" text in the language-independent specification description for the status parameter - it does not affect parameter names in bindings or in text. Please actually look at page 38 (printed) line 2 in the mpi-4-rc PDF to see the "problem".

**Comment by wgropp** [comment]
On the line breaking problems.  This breaks for the same reason the Watson decided Toronto was in the US - there is a linear penalty function used to decide (in the case of TeX, which collection of line breaks in a paragraph will minimize the cost function). This can be fixed with a combination of "short" versions and, as noted, re-ordering some of the text (apparently, this used to be common, even in literature!).

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@wgropp do you have a specific recommendation for this paragraph? If not, I think we should leave this alone --- it is not wrong and there are other suggestions that take priority.

**Comment by dholmes-epcc-ed-ac-uk** [comment]
"array_of_status -> array_of_statuses" fixed by https://github.com/mpi-forum/mpi-standard/pull/363

**Comment by dholmes-epcc-ed-ac-uk** [comment]
"missing Oxford comma" fixed by https://github.com/mpi-forum/mpi-standard/pull/364

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@GuillaumeMercier 
> Adding "and binds to it all arguments of a send operation." to the description of MPI_[B|S|R]send_init to be consistent with MPI_Send_init?

I would prefer to got the other way and remove that sub-clause from MPI_SEND_INIT.

**Comment by dholmes-epcc-ed-ac-uk** [comment]
"and binds to it" fixed by https://github.com/mpi-forum/mpi-standard/pull/365

**Comment by dholmes-epcc-ed-ac-uk** [comment]
"send with ready mode" fixed by https://github.com/mpi-forum/mpi-standard/pull/366

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@GuillaumeMercier I believe the commas before&after \mpiarg{request} on p95 are correct --- leaving alone.

**Comment by GuillaumeMercier** [comment]
> @GuillaumeMercier I believe the commas before&after \mpiarg{request} on p95 are correct --- leaving alone.

On second thoughts, I think so too.

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@GuillaumeMercier I have chosen to do a much larger change instead of just "Start" -> "Starts" for MPI_Start_all.

I have replaced the entire paragraph with wording copied&pasted&tweaked from the new MPI-4.0 wording for MPI_WAITALL:

> The error-free execution of MPI_STARTALL has the same effect as the execution of MPI_START for each of the array elements in some arbitrary order. MPI_STARTALL with an array of length one is equivalent to MPI_START.

This replaces the inline code with text - this was missed in the previous ticket that did this text change for MPI_WAITALL (and others). The replacement obviates the need for your one-letter fix.

**Comment by dholmes-epcc-ed-ac-uk** [comment]
"Start ->Starts" obviated by https://github.com/mpi-forum/mpi-standard/pull/368 ** PR needs CC check **

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@GuillaumeMercier I'm not adding "associated" --- the rest of that paragraph refers to the request itself changing state, which is correct, IMHO. It raises questions of: associated with what? Is it not the request that is passed into this start procedure?

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@regrant 

> These point-to-point persistent calls involve no communication.
>  -> These point-to-point persistent procedures are local.

Oh yes, I'd love to make that change to the semantics but no, that is not a CC level change.

Semantic Terms WG: "involve no communication" is *much* stronger than "are local" (weak local permits dependency on remote progress, which --- in some MPI implementations --- requires remote MPI procedure calls, i.e. arbitrary delay controlled by the execution state of at least one remote process).

This is at least errata but probably a full proposal - and is unnecessary because partitioned point-to-point with one partition gives the semantic that persistent point-to-point should have had since forever. We should maintain the existing API until we can deprecate it in favour of MPI_Psend.

**Comment by wgropp** [comment]
On the line breaking, no, while this is simple to fix, its not a priority, and it does require inserting macros and/or changing explicit line breaking penalties.

**Comment by dholmes-epcc-ed-ac-uk** [comment]
"spare commas around then" fixed by https://github.com/mpi-forum/mpi-standard/pull/369

**Comment by dholmes-epcc-ed-ac-uk** [comment]
"persistent collective word ordering" fixed by https://github.com/mpi-forum/mpi-standard/pull/371

**Comment by dholmes-epcc-ed-ac-uk** [comment]
"Additional cross-refs" fixed by https://github.com/mpi-forum/mpi-standard/pull/372
"Confusing line break" fixed by https://github.com/mpi-forum/mpi-standard/pull/372

**Comment by dholmes-epcc-ed-ac-uk** [comment]
"Take account of persistent collectives etc for MPI_WAIT" fixed by https://github.com/mpi-forum/mpi-standard/pull/373 ** PR needs CC check **

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@devreal & @raffenet in response to your p109 change "what about other request-based things for MPI_WAIT" I have modified the "sends and receives" to "any request-based nonblocking or persistent operation".

request-based: MPI_PUT is nonblocking but cannot be completed with MPI_WAIT.

**Comment by dholmes-epcc-ed-ac-uk** [comment]
I'm done for now.

Unaddressed issues: Status vs status and bad line-breaks.

**Comment by wesbland** [comment]
Comments from @bkmgit:

- [ ] Section 3.5 Order pg. 51
 * On many networks, there are multiple paths between source and destination and the routing algorithm may choose different paths for two messages. It may be the case that the second message arrives before the first one. This may happen even for single threaded execution. One might suggest it best to use tags to allow for this which may be a helpful feature for asynchronous computations. If different tags are not used, then default behavior should be fine. Probably, this would be a change for future versions of MPI.

- [ ] Examp 3.8 pg. 54 line 10 - "The receive operation .." could perhaps be "Both ranks call receive operation first, so .."

- [ ] Example 3.9 pg. 54 line 26 - "The message sent by each process .." could perhaps be, "Both ranks send first, so the message .."

- [ ] May be helpful to have an explanation of why MPI_COUNT_KIND is used. Also, if this is only used for count variables, maybe the display of Fortran 2008 interfaces could be done only once, with an explanation added for count being able to be declared as either integer or MPI_COUNT_KIND

- [ ] MPI_Isendrecv pg. 65 - This is missing Fortran 2008 bindings with INTEGER(KIND=MPI_COUNT_KIND)

- [ ] MPI_Isendrecv_replace pg. 66 - This is missing Fortran 2008 bindings with INTEGER(KIND=MPI_COUNT_KIND)

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@bkmgit 

§3.5 Order: I do not know what change you are suggesting here. However, I am reticent to make any change to the semantics of point-to-point communication without a full proposal and lots of time for scrutiny by the whole MPI community.

Ex3.8: the rest of the sentence makes less sense with your suggested change than without. The requirement that the receive must complete before the send at the first process is true irrespective of the ordering of the code at the other process. Adding extraneous information to the condition does not make this clearer.

Ex3.9: similarly, the requirement that the message must be copied out before the send operation completes is not a consequence of the ordering of code on the other process.

There is an explanation of when/why MPI_COUNT_KIND is used - it is chapter 2.

MPI_Isendrecv has not been Embiggened (tm): good catch, thanks!
MPI_Isendrecv_replace has not been Embiggened (tm): good catch, thanks!

**Comment by dholmes-epcc-ed-ac-uk** [comment]
"Missing Embiggening for nonblocking send-recv procedures" fixed by https://github.com/mpi-forum/mpi-standard/pull/384

**Comment by bkmgit** [comment]
Ok on 3.5 - more an observation that guaranteeing order of messages may not be necessary for algorithms that use asynchronous communication. It would take time to make such changes if they would be helpful.

For examples 3.8 and 3.9, had to read through the code several times to determine the differences, though if someone is reading more than one section, they will likely be able to spot the differences.

Ok on MPI_COUNT_KIND 

Thanks for the changes.

**Comment by gpapaure** [comment]
§ 3.5 : Embiggened MPI_Sendrecv and MPI_Sendrecv_replace are missing (pg80, 81, 82)
Making the PR

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@gpapaure 

>"Missing Embiggening for nonblocking send-recv procedures" fixed by mpi-forum/mpi-standard#384

**Comment by gpapaure** [comment]
This PR seems to be for non-blocking send-recv calls.
Blocking one are missing too, I can do a new PR or you can complete the existing PR for more consistency

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@gpapaure oh wow - no idea how that got missed - good catch!

I'll update PR 384 to fix this. Thanks 👍

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@gpapaure should be done now.

**Comment by gpapaure** [comment]
perfect. Thanks

**Comment by dholmes-epcc-ed-ac-uk** [comment]
"Case of Status in LIS should be status" fixed by https://github.com/mpi-forum/mpi-standard/pull/390

**Comment by dholmes-epcc-ed-ac-uk** [comment]
All points addressed by PRs and all linked PRs are merged. Closing this issue.
